internal/testing/events/recorder.go

```diff
@@ -0,0 +1,110 @@
+ limitations under the License.
+*/
+
+package events
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"gotest.tools/v3/assert"
+	corev1 "k8s.io/api/core/v1"
+	eventsv1 "k8s.io/api/events/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/tools/record/util"
+	"k8s.io/client-go/tools/reference"
+)
+
+// Recorder implements the interface for the deprecated v1.Event API.
+// The zero value discards events.
+// - https://pkg.go.dev/k8s.io/client-go@v0.24.1/tools/record#EventRecorder
+type Recorder struct {
+	Events []eventsv1.Event
+
+	// eventf signature is intended to match the recorder for the events/v1 API.
+	// - https://pkg.go.dev/k8s.io/client-go@v0.24.1/tools/events#EventRecorder
+	eventf func(regarding, related runtime.Object, eventtype, reason, action, note string, args ...interface{})
+}
+
+// NewRecorder returns an EventRecorder for the deprecated v1.Event API.
+func NewRecorder(t testing.TB, scheme *runtime.Scheme) *Recorder {
+	t.Helper()
+
+	var recorder Recorder
+
+	// Construct an events/v1.Event and store it. This is a copy of the upstream
+	// implementation except that t.Error is called rather than klog.
+	// - https://releases.k8s.io/v1.24.1/staging/src/k8s.io/client-go/tools/events/event_recorder.go#L43-L92
+	recorder.eventf = func(regarding, related runtime.Object, eventtype, reason, action, note string, args ...interface{}) {
+		t.Helper()
+
+		timestamp := metav1.MicroTime{Time: time.Now()}
+		message := fmt.Sprintf(note, args...)
+
+		refRegarding, err := reference.GetReference(scheme, regarding)
+		assert.Check(t, err, "Could not construct reference to: '%#v'", regarding)
+
+		var refRelated *corev1.ObjectReference
+		if related != nil {
+			refRelated, err = reference.GetReference(scheme, related)
+			assert.Check(t, err, "Could not construct reference to: '%#v'", related)
+		}
+
+		assert.Check(t, util.ValidateEventType(eventtype), "Unsupported event type: '%v'", eventtype)
+
+		namespace := refRegarding.Namespace
+		if namespace == "" {
+			namespace = metav1.NamespaceDefault
+		}
+
+		recorder.Events = append(recorder.Events, eventsv1.Event{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      fmt.Sprintf("%v.%x", refRegarding.Name, timestamp.UnixNano()),
+				Namespace: namespace,
+			},
+			EventTime:           timestamp,
+			Series:              nil,
+			ReportingController: t.Name(),
+			ReportingInstance:   t.Name() + "-{hostname}",
+			Action:              action,
+			Reason:              reason,
+			Regarding:           *refRegarding,
+			Related:             refRelated,
+			Note:                message,
+			Type:                eventtype,
+		})
+	}
+
+	return &recorder
+}
+
+var _ record.EventRecorder = (*Recorder)(nil)
+
+func (*Recorder) AnnotatedEventf(object runtime.Object, annotations map[string]string, eventtype, reason, messageFmt string, args ...interface{}) {
```

Comment:
:thinking: Why do we need this method if it results in a panic? Is it just to match the [record package](https://pkg.go.dev/k8s.io/client-go/tools/record#FakeRecorder.AnnotatedEventf)?

Reply:
Yep. [EventRecorder](https://pkg.go.dev/k8s.io/client-go/tools/record#EventRecorder) is the interface from the record package, and it's what [controller-runtime uses](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.12.1/pkg/cluster#Cluster).

The [future](https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/383-new-event-api-ga-graduation) is the [events package](https://pkg.go.dev/k8s.io/client-go/tools/events#EventRecorder), which has a simpler API (and implementation!). That future has no "annotated event" concept, so we should stay away from it now.